### ISSUE TITLE
Accept vs Response-Content-Type

### ISSUE THREAD

**Issue by Guergeiro** [drashland/deno-drash#173]
## Discussion
Why use a custom `Response-Content-Type` header instead of the one already defined and standardized ([Accept](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.1))?

**Comment by crookse** [comment]
You bring up a very good point actually. So like I said in the YouTube chat, the `Response-Content-Type` header tells the server exactly what content type the user wants. I know `Accept` does the same, but it's not as explicit as they use the words, "We 'prefer' this content type, but any of the following are fine as well."

https://drash.land/docs/#/advanced-tutorials/content-negotiation/user-profiles/part-4 has a bit more on the `Response-Content-Type` header.

But like I said, you bring up a very good point. During my development for serving a specific content type and setting the `Content-Type` header on the response, I realized `Accept` isn't as explicit because it's a list of content types as opposed to just one. Do you think it's ok that the response object uses the `Accept` header and chooses the first content type in the list and sets that as the response's `Content-Type`?

It's really open for discussion. I'm all for using the `Accept` header, but I just don't know how to make it more explicit when it comes to Drash. Any ideas you have are greatly appreciated!

**Comment by crookse** [comment]
This is an example of how the `Accept` header could be used as of right now:

```typescript
import Drash from "https://deno.land/x/drash@v0.39.5/mod.ts";

class HomeResource extends Drash.Http.Resource {

  static paths = ["/"];

  public GET() {
    this.response.body = false;

    let contentTypes = this.request.headers.get("Accept").split(";");

    contentTypes.forEach((contentType: string) => {
      if (!this.response.body) {
        switch (contentType) {
          case "text/html":
            this.response.headers.set("Content-Type", contentType);
            this.response.body = this.htmlResponse();
            break;
          case "application/json":
            this.response.headers.set("Content-Type", contentType);
            this.response.body = this.jsonResponse();
            break;
          case "text/xml":
            this.response.headers.set("Content-Type", contentType);
            this.response.body = this.xmlResponse();
            break;
        }
      }
    });

    if (!this.response.body) {
      this.response.body = "You requested a content-type that isn't supported by this resource.";
    }

    return this.response;
  }

  protected jsonResponse() {
    return {hello: "world"};
  }

  protected htmlResponse() {
    return "<div>Hello, world!</div>";
  }

  protected xmlResponse() {
    return "<body>Hello, world!</body>";
  }
}

const server = new Drash.Http.Server({
  address: "localhost:1447",
  response_output: "text/html",
  resources: [HomeResource]
});

server.run();
```

```

curl -X GET -H "Accept: application/json;text/html" localhost:1447
{"hello":"world"}

curl -X GET -H "Accept: text/html;application/json" localhost:1447
<div>Hello, world!</div>

curl -X GET -H "Accept: text/xml;application/json" localhost:1447
<body>Hello, world!</body>

```

**Comment by Guergeiro** [comment]
In my opinion, Drash should stick to web (browser?) standards as much as possible (we are using Deno after all), even if takes a bit of stretching from a framework maintainer point of view. And since I believe this framework will be big when Deno starts to pick up, I think all these **core** components/design choices must be extensively looked over.

The first thing I envisioned was exactly what you said, a *precedence approach* and all the subsequent being fallback in case the server doesn't understand it. Until you reach a point that you either found one that satisfies the server or just spit out a `5xx` code of any sort.

**Comment by crookse** [comment]
I understand. I think I'm going to make it a priority to get this breaking change out before Drash v1.0.

Thanks, @Guergeiro! Really appreciate your thoughts!

**Comment by Guergeiro** [comment]
> This is an example of how the `Accept` header could be used as of right now:

That's a really good solution in my opinion. Maybe, instead needing the user to implement that, already have it done under the hood and maybe give an array with all options already parsed to the user.

**Comment by Guergeiro** [comment]
From https://github.com/drashland/deno-drash/issues/173#issuecomment-613590632

> `5xx` code of any sort.

Actually a `415` might be better.

**Comment by crookse** [comment]
I like the 415. It seems to be the correct choice. The user supplied an unsupported content type, therefore it's a client error.

**Comment by ebebbington** [comment]
After some brainstorming with @crookse, we may have a basic, workable and clean solution (in theory and still in the planning stage). This is all very early so allow for minor details here and there but very open to discussion and suggestions.

**Drash's Approach**
Here is what we think Drash would need to implement, to move over to the `Accept` header

1. A new `accepts` method on `Drash.Http.Request`
```
public accepts (type: string|string[]): boolean
```

2. This method will be public and will accept a string or an array of strings, such as "text/html" or any content type.

3. This method will then check if that content type is apart of the accept header, for example see the pseudo code below. I'm skeptic about passing in an array... because you could do `accepts(['html, json'])` and not know which one is accepted...
```
public accepts(type: string|string[]): boolean {
    // Fixme :: The below does not account for q=0.5
    let firstAcceptHeader = this.request.headers.get("accept").split(";")[0];
    // or this
    let acceptHeaders= this.request.headers.get("accept").split(";");

    // Now check if the request will accept the content type asked for
    if (typeof type === 'string') return acceptHeaders.indexOf(type) >= 0
  
    // else if its an array of content types, for example choose html, if it doesnt have html then choose json
    // So it will prioritise the content types from left to right passed in
    const match = type.forEach(t => acceptHeaders.indexOf(t) >= 0).length >= 1
    return match  
```

I think that would be it for the Drash implementation

**User Implementation**
So this would have to be documented in the docs, and the user would interact with the `accepts` method. So take the following resource as an example:
```
public GET() {
    // this is a JSON-api like endpoint so only send back json
    if (this.request.accepts("application/json")) {
      return JSON.stringify({
          some: "data"
      })
    } else {
      throw 400 error or something // needs more work
    }
}
```

**Client Implementation**
By Client, it's basic HTTP requests, whether a page is requested or a JSON fetch.

So obviously most requests will be `any` or "text/plain", but i want to focus on the fetch side of things, s say you have some api-like endpoints, the user could *strictly* declare what *they expect from the server* and using the above, the server would respond accordingly:
```
fetch(some url, {
  headers: {
    'Accept': 'application/json'
  }
})
```

**In a Nutshell**
A user of Drash would specifically check if the accepted content type matches te correct request, which i feel makes sense, as you might have an endpoint that shows html, so you only want to display it is the accept header accepts html.. i think?

**Comment by Guergeiro** [comment]
From https://github.com/drashland/deno-drash/issues/173#issuecomment-613630617

> 3. This method will then check if that content type is apart of the accept header, for example see the pseudo code below. I'm skeptic about passing in an array... because you could do `accepts(['html, json'])` and not know which one is accepted...

I don't find it bad in anyway. The developer needs to understand that, if they want to join implementations, they can pass an *array* and implement the functionality. Otherwise, just be explicit and do it one by one.

> ```typescript
> const acceptHeaders = this.request.headers.get("accept").split(";");
> ```

Here I think an *regex* before `split(";")` to remove in between *spaces* is required (not sure though). Different clients might request the endpoint with different `Accept` headers:
```json
"Accept": "application/json;text/html"
```
```json
"Accept": "application/json; text/html"
```
```json
"Accept": " application/json; text/html"
```

I would have the `public accepts (type: string|string[]): boolean` like this:
```typescript
public accepts (type: string|string[]): boolean => {
    const acceptHeaders = this.request.headers.get("accept").replace(/\s/g, "").split(";");
    if (typeof string == "string") {
        return acceptHeaders.indexOf(type) != -1;
    }
    return type.some((element) => acceptHeaders.indexOf(element) != -1);
};
```

**Comment by crookse** [comment]
I still can't wrap my head around what the client would actually want first unless we have a `request.acceptsFirst()` or something. Let's say the client passes in the following:

```
Accept: text/html;application/json;text/xml
```

If in the resource `this.request.accepts("application/json")` is called, then that would resolve to `true`, and the resource would return a JSON response; but the first acceptable content type is `text/html`.

**Comment by Guergeiro** [comment]
From https://github.com/drashland/deno-drash/issues/173#issuecomment-613682348

Well, then we need to invert the dependencies... Using your [example](https://github.com/drashland/deno-drash/issues/173#issuecomment-613589659) might work.
```typescript
public GET() {
    const contentTypes: string[] = someAwesomeFunctionThatAlreadyDoesThatForMe();

    // We immediately return at the first one found! By order of the client
    for (const content of contentTypes) {
        switch(content) {
            case "application/json":
                 this.response.set("Content-Type", "application/json");
                 this.response.body = this.jsonRespone();
                 return this.response;
            case "text/html":
                 this.response.set("Content-Type", "text/html");
                 this.response.body = this.htmlRespone();
                 return this.response;
        }
    }
    // If we arrive at this point, means that no content type is satisfied
    this.response.set("Content-Type", contentTypes[0]);
    this.response.status(415);
    this.response.body("Server doesn't understand a thing :)");
    return this.response;
}
protected jsonResponse() {
    return {hello: "world"};
}
protected htmlResponse() {
    return "<div>Hello, world!</div>";
}
```

**Comment by Guergeiro** [comment]
Now this might be a bit hardcoded and hard to maintain in really big projects (but maybe that's what *A REST microframework for Deno* is).

A solution to this could be some sort of implementation using TypeScript decorators (just giving some random thoughts).
```typescript
@accept("application/json")
public GET() {
    // this.response.set("Content-Type", "application/json"); -> not need since decorator should handle that
    this.response.status(200);
    this.response.body(JSON.stringify({hello: "world"}));
    return this.response;
}
@accept("text/html")
public GET() {
    // this.response.set("Content-Type", "text/html"); -> not need since decorator should handle that
    this.response.status(200);
    this.response.body("<body>Hello World!</body>");
    return this.response;
}
```
Something like this. Don't know, but seems clean.

**Comment by crookse** [comment]
I like the first approach better. I'm picky about Drash's syntax and decorators have been a no since day 1. You're right though. **_Return at the first one found._**

@ebebbington , you got that? ;)

**Comment by crookse** [comment]
Btw thanks for your help on this @Guergeiro

**Comment by Guergeiro** [comment]
From https://github.com/drashland/deno-drash/issues/173#issuecomment-613714909

> decorators have been a no since day 1

Do you have a reason for that? Because you might need to consider them in the future (depending how Drash evolves of course).

> ***Return at the first one found.***

The problem with this approach is that you delegate the work to the developer, whereas with *decorators* that could've been done by the framework. But again, if you look at ExpressJS from Node for example, you as developer do it.

**Comment by Guergeiro** [comment]
From https://github.com/drashland/deno-drash/issues/173#issuecomment-613716296
> Btw thanks for your help on this @Guergeiro

No worries. I won't PR anything because Deno is not v1.0 yet. In the future, I'm looking forward to help more intensively. For now, I'll just lurk in the issues/docs 😉

**Comment by ebebbington** [comment]
@crookse @Guergeiro So would we want the `accepts` method to return the matched type or a boolean? Guess this really only affects the *developer*, and which one would be better suited

Assume the request only accepts application/json:
```
const isAccepted = this.request.accepts(["text/html", "application/json"]); // true
if (isAccepted) ...
// or
const acceptedType = this.request.accepts(["text/html", "application/json"]); // "application/json"
if (acceptedType === "text/html") ...
if (acceptedType === "application/json") ...
```

**Comment by crookse** [comment]
I say the matched type

**Comment by ebebbington** [comment]
Makes sense to be honest, i might think of returning false then if none were found as we will be returning the content type of data type `string`

**Comment by Guergeiro** [comment]
Matched type would be better. And as `return` value, I wouldn't mind a `null`. It will fall under an `else` statement either way.
